### PR TITLE
client: file and environment variables based configuration

### PR DESCRIPTION
**Pull request by mitake**
Current etcd client library doesn't provide a mechanism for creating
client.Client object based on file or env vars. Therefore users must
write their own feature for the purpose in their applications.

For avoiding the duplicated implementations, this PR adds new
functions for creating client with file and environment variables
based on viper.

NewWithFile(client.Config, configPath string): create a new Client
with a file. Any file formats supported by viper will be accepted.

NewWithEnv(client.Config): create a new Client with values in
environment variables.

Like the existing New(), application must pass a client.Config object
because there are some fields which isn't related to configuration
(e.g. Transport).

Fixes https://github.com/coreos/etcd/issues/4008

### CODE REVIEW

**Comment by mitake** [comment]
godep seems to be failed... Could you share your best practice of adding new dependencies with godep? @xiang90 @gyuho 

I tried for a while but I couldn't understand how to add new dependencies successfully... :( Path rewriting seems to be failed in both cases of -r or without -r.

**Comment by gyuho** [comment]
Godep save -r -t ./...

On Wed, Jan 6, 2016, 04:18 Hitoshi Mitake <notifications@github.com> wrote:

> godep seems to be failed... Could you share your best practice of adding
> new dependencies with godep? @xiang90 <https://github.com/xiang90> @gyuho
> <https://github.com/gyuho>
>
> I tried for a while but I couldn't understand how to add new dependencies
> successfully... :( Path rewriting seems to be failed in both cases of -r or
> without -r.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/4144#issuecomment-169310762>.
>
-- 
Sincerely,
Gyu-Ho Lee

**Comment by gyuho** [comment]
I am not familiar with viper. Could you provide simple input/output from this feature?

**Comment by mitake** [comment]
@gyuho thanks for your sharing. Do you `godep save -r -t ./...`? not `godep save -r -t github.com/coreos/etcd`?

At least viper supports toml, json, and yaml.

**Comment by gyuho** [comment]
I do `godep restore` first if my machine does have all the dependencies from etcd Godep.
And you run `godep save -r -t ./...` INSIDE etcd directory and in your branch.  This will add
missing dependencies. This is just one way to do it. There should be other ways to do it.

**Comment by mitake** [comment]
Yes, I executed `godep restore` before. After that, `godep save -r -t ./...` reports below error:
```
godep: Package (github.com/godbus/dbus) not found
```

It seems that github.com/coreos/go-systemd depends on the above dbus package. But the package cannot be found in Godeps (therefore `godep restore` cannot save it in gopath) so `godep save` would be reporting error... Do you have any ideas on this problem?

**Comment by gyuho** [comment]
```
go get -v -u github.com/coreos/go-systemd
go get -v -u github.com/godbus/dbus
```

And try again?

**Comment by mitake** [comment]
Yes, I tried to get the missing packages (after getting dbus, `godep save` complained about other missing packages repeatedly (e.g. github.com/stretchr/objx, golang.org/x/text/encoding, etc,etc...).

Finally, `godep save` saied it needs google.golang.org/api/container/v1beta1. And the package is not provided already (newer version google.golang.org/api/container/v1 seems to be provided). So I couldn't get it for satisfying dependency :(

BTW, godep warns like below:
```
godep: WARNING: Recorded go version (go1.5.1) with minor version string found.
godep: To record current major go version run `godep update -goversion`.
```
do you think that this warning is related to the problem?

**Comment by gyuho** [comment]
Yeah Godep can sometimes be total mess
Try to add only the dependency that you need and
see if it builds. If it does, just commit your change.

On Wed, Jan 6, 2016, 17:29 Hitoshi Mitake <notifications@github.com> wrote:

> Yes, I tried to get the missing packages (after getting dbus, godep save
> complained about other missing packages repeatedly (e.g.
> github.com/stretchr/objx, golang.org/x/text/encoding, etc,etc...).
>
> Finally, godep save saied it needs google.golang.org/api/container/v1beta1.
> And the package is not provided already (newer version
> google.golang.org/api/container/v1 seems to be provided). So I couldn't
> get it for satisfying dependency :(
>
> BTW, godep warns like below:
>
> godep: WARNING: Recorded go version (go1.5.1) with minor version string found.
> godep: To record current major go version run `godep update -goversion`.
>
> do you think that this warning is related to the problem?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/4144#issuecomment-169517788>.
>
-- 
Sincerely,
Gyu-Ho Lee

**Comment by mitake** [comment]
Seems that the oddly required packages are used by commands and not used features in the truly required packages... Is it okay to remove the unused code manually?

**Comment by gyuho** [comment]
No we should include all files that Godep tries to grab. There are too many
edge cases to do that.

On Wed, Jan 6, 2016, 17:52 Hitoshi Mitake <notifications@github.com> wrote:

> Seems that the oddly required packages are used by commands and not used
> features in the truly required packages... Is it okay to remove the unused
> code manually?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/4144#issuecomment-169521157>.
>
-- 
Sincerely,
Gyu-Ho Lee

**Comment by mitake** [comment]
Then the simplest solution would be implement etcd client own config mechanism based on viper. Is it acceptable? @gyuho @xiang90 

Of course if there is a better library, using it would be an easier way. But I think viper is suitable for configuration of the etcd client library because it supports env vars and files in a unified manner and supports watching file update. Do you know any other good candidates?

**Comment by gyuho** [comment]
Yeah https://github.com/coreos/etcd/issues/4008 does not specify what kind of design we need.
Please discuss with us if you proceed any further. I don't have strong preference to library but viper
seems like a great choice.

Defer to @xiang90

**Comment by mitake** [comment]
I noticed that even on master branch, godep restore and godep save -r ./... fails with this message:
```
godep: Package (github.com/godbus/dbus) not found
```

It seems very odd... maybe used part of go-systemd wouldn't depending on godbus, but unused part would depend on godbus?

How should I solve it? :(

**Comment by gyuho** [comment]
`github.com/godbus/dbus` is in your local machine? If not, it wouldn't work.
Last time I added dependency (spf13/cobra), it worked fine.

If you think this is godep problem, can you file an issue to godep?

But godep should work. One way, just try to remove all your local packages in your 
GOPATH, and git clone etcd. And godep restore to copy all the dependencies.
And do your patch. And try again with godep save -r -t ./...

This should work.

**Comment by gyuho** [comment]
@mitake Did you figure out the godep issue?
And do you think we need the whole `viper` project code-base to achieve this?

Thanks,

**Comment by mitake** [comment]
@gyuho I'm still digging it but seems that importing viper isn't the root cause of the problem. Importing viper from a new clean slate project doesn't cause any problems.

The problem would come from godep or Godeps/ directory content of etcd. After solving it, adding viper to Godeps/ would work fine. I'll update this PR ASAP.

**Comment by gyuho** [comment]
@mitake Ok, let me know if you need any help.

Thanks!

**Comment by mitake** [comment]
@gyuho could you share your godep's version? seems that recent commits of godep changed its downloading behavior.

**Comment by gyuho** [comment]
`godep v45 (linux/amd64/go1.5.2)` but I think it shouldn't matter?

**Comment by mitake** [comment]
@gyuho thanks for sharing the version.

Could you also share the content of your `$GOPATH/src/github.com/coreos/go-systemd`? It seems that godep obtains unused packages e.g. `github.com/coreos/go-systemd/dbus` and they would be the source of the problem. Though I'm not fully sure yet...

**Comment by gyuho** [comment]
@mitake 

Just tried to `godep` viper myself:

```go
package main

import (
	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/spf13/viper"
	"github.com/coreos/etcd/etcdmain"
)

func main() {
	etcdmain.Main()
	_ = viper.GetStringSlice
}

```

It works for me, with no problem:

```
On branch master
Your branch is ahead of 'origin/master' by 749 commits.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   Godeps/Godeps.json
	modified:   Godeps/_workspace/src/github.com/spf13/cobra/cobra/cmd/add.go
	modified:   Godeps/_workspace/src/github.com/spf13/cobra/cobra/cmd/helpers.go
	modified:   Godeps/_workspace/src/github.com/spf13/cobra/cobra/cmd/init.go
	modified:   Godeps/_workspace/src/github.com/spf13/cobra/cobra/cmd/root.go
	modified:   main.go

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	Godeps/_workspace/src/github.com/BurntSushi/
	Godeps/_workspace/src/github.com/hashicorp/
	Godeps/_workspace/src/github.com/kr/pretty/
	Godeps/_workspace/src/github.com/kr/text/
	Godeps/_workspace/src/github.com/magiconair/
	Godeps/_workspace/src/github.com/mitchellh/
	Godeps/_workspace/src/github.com/spf13/cast/
	Godeps/_workspace/src/github.com/spf13/jwalterweatherman/
	Godeps/_workspace/src/github.com/spf13/viper/
	Godeps/_workspace/src/gopkg.in/

no changes added to commit (use "git add" and/or "git commit -a")

```

**Comment by gyuho** [comment]
`github.com/coreos/go-systemd` has only one external dependency `github.com/godbus/dbus`. I don't know why you cannot `godep` it. Can you file an issue to `godep`?

**Comment by mitake** [comment]
@gyuho could you share your every command line for getting and saving viper? I'd like to follow your way, it it fails, I'll file an issue to godep. I'm very new to godep so I would be missing something...

**Comment by gyuho** [comment]
@mitake 

1. just clone the fresh etcd
2. `go get -v -u github.com/tools/godep`
3. edit main.go in etcd directory, as above, but with `github.com/spf13/vipe` as the import path
4. `godep save -r -t ./...` in etcd directory
5. `github.com/spf13/vipe` will become `github.com/coreos/etcd/Godeps/_workspace/src/github.com/spf13/viper`

**Comment by mitake** [comment]
@gyuho thanks a lot! BTW, is the 2nd line
`go get -v -u github.com/spf13/viper` ?

**Comment by gyuho** [comment]
second line is optional if you already have `godep` and `viper`. It just pulls down the latest code.

**Comment by mitake** [comment]
In addition, for doing `go get`, `GOPATH` must be set. Which directory are you using as `GOPATH`? e.g. ~/gopath, etcd.git/gopath?

**Comment by gyuho** [comment]
`etcd.git/gopath` is only used when building `etcd` in my machine.
I put all go code under `$HOME/go` (`$GOPATH`)

**Comment by mitake** [comment]
I followed your command lines. After 4th line, the etcd repository was changed like attached status.
[git-status.txt](https://github.com/coreos/etcd/files/91436/git-status.txt)

Have you ever seen such status?

**Comment by gyuho** [comment]
I don't think you did it right. Did you run `godep save -r -t ./...` in `$GOPATH/src/github.com/coreos/etcd`?

**Comment by gyuho** [comment]
@mitake If it doesn't work, don't worry about it.

Just remove the godep commit, and leave `viper` as an external dependency.
I can take care of dependency later.

Just try to get the feature approved in this PR.

**Comment by mitake** [comment]
No, I just executed `godep save -r -t ./...` at the top directory of cloned etcd.

BTW, like your instruction, I'm setting `GOPATH` as `~/gopath`. So the directory doesn't have `src/github.com/coreos/etcd`. When did you created it? Should I create it manually?

**Comment by gyuho** [comment]
I don't use `YOUR_ETCD_PATH/gopath` because it only creates when you run `build` script.
All I do is have `etcd` code under `$GOPATH/src/github.com/coreos`. There could be other ways
but this works for me.

**Comment by mitake** [comment]
Then, could you add viper? Really sorry for that... I'll update this PR soon.

**Comment by mitake** [comment]
Updated, could you push the viper import commit in your repository? Then I can cherry-pick your commit and this PR will be CI ready. Sorry for bothering :(

**Comment by gyuho** [comment]
Np.

Yeah let's just do without godep. We can do godep later.
Can you also add tests or example?

**Comment by mitake** [comment]
ok, I'll add tests.

**Comment by mitake** [comment]
@gyuho added tests, could you take a look?

File: client/client_test.go [hunk 1]
```diff
@@ -894,3 +895,132 @@ func TestHTTPClusterClientResetPinRandom(t *testing.T) {
...
+		}
+
+		cfg := Config{}
+		result, cerr := NewWithEnv(cfg)
+		if cerr != nil {
+			t.Fatalf("#%d: test failed (%v)", i, cerr)
+		}
+
+		resultEps := result.Endpoints()
+		expectedEps := tt.expected.Endpoints()
```
Comment (gyuho) [inline client/client_test.go]: Can we compare with `reflect.DeepEqual`?

**Comment by mitake** [comment]
@gyuho updated for using `reflect.DeepEqual`, PTAL.

**Comment by gyuho** [comment]
@mitake Thanks,

This weekend, I will try to play with this to get better idea of what it is doing.

File: client/client_test.go [hunk 2]
```diff
@@ -894,3 +895,127 @@ func TestHTTPClusterClientResetPinRandom(t *testing.T) {
...
+		},
+	}
+
+	for i, tt := range tests {
+		path := "test." + tt.ext
+		f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE, 0666)
+		if err != nil {
+			t.Fatalf("#%d creating config file for test failed (%v)", i, err)
+		}
+		defer os.Remove(path)
```
Comment (gyuho) [inline client/client_test.go]: `os.Remove` at the end of each iteration without `defer`

File: client/client.go [hunk 3]
```diff
@@ -205,6 +208,86 @@ func New(cfg Config) (Client, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	ext := filepath.Ext(configPath)
+	if len(ext) != 0 {
+		viper.SetConfigType(ext[1:])
+	}
+
+	err = viper.ReadConfig(f)
```
Comment (gyuho) [inline client/client.go]: So does this config set the viper object in global scope, if I understand correct?

**Comment by gyuho** [comment]
According to https://github.com/spf13/viper#why-viper, viper prefers the top most, in the order below:

```
explicit call to Set
flag
env
config
key/value store
default
```

So as long as we leave `viper` as an option to overwrite some environment variables,
the user flags without using `viper` won't be affected. I tried it and it worked. And I see
environment variable prefix setting correctly `ETCDCLIENT_` (this env var is not used
anyway).

Even though this is for very limited use case, if we start using this, we need some
documentation on this as well. Defer to @xiang90 for his final decision.

File: client/client_test.go (hunk 2 above)
Comment (mitake) [inline client/client_test.go]: Ah, sorry for that. I'll fix it soon.

File: client/client.go (hunk 3 above)
Comment (mitake) [inline client/client.go]: Yes, it would be a common usage of viper without multiple configuration sources.

**Comment by mitake** [comment]
@gyuho ok, I'll prepare docs tomorrow. Please wait for a while.

**Comment by mitake** [comment]
Kindly ping, @xiang90 . How do you think about this idea? I believe this feature is useful. Actually some users (including k8s) seem to be facing the lack of this kind of feature: https://github.com/coreos/etcd/issues/4008 .

P.S. I'm planning to add the doc to the file introduced in this PR: https://github.com/coreos/etcd/pull/4231

**Comment by philips** [comment]
It would be nice to see some fixtures and examples using this code in the commit. Also, some human readable doc describing the schema.

**Comment by mitake** [comment]
@philips thanks for your comment. I'll add the stuff later.

@gyuho could you push your commit for importing viper to somewhere public? I'd like to cherry-pick the commit and let this PR have it. Sorry for the time, I'm still not familiar with godep... :(

**Comment by gyuho** [comment]
@mitake Don't worry about `godep` now. Once this gets approved, I can cherry-pick your commit
and merge with `godep` added.

You can go ahead addressing Brandon's feedback.

@xiang90 is currently not available. Let's wait for him to come back, which is next week.

Sorry for delay.

**Comment by mitake** [comment]
@philips @gyuho I added an example under the directory contrib/client_conf_example and a doc to client/README.md. Could you take a look?

**Comment by mitake** [comment]
@gyuho now importing viper with godep seems to be successful

**Comment by mitake** [comment]
Kindly ping, @philips @gyuho ?

**Comment by xiang90** [comment]
@mitake I will take a look early next week. I am not sure how this will work with etcdctl, since etcdctl also takes env_var.

**Comment by mitake** [comment]
@xiang90 Thanks. etcdctl isn't affected because it doesn't use newly added functions `NewWithEnv()` or `NewWithFile()`.

**Comment by xiang90** [comment]
@mitake But we also want etcdctl to take env_var in the future. So I feel they conflict with each other.

**Comment by mitake** [comment]
@xiang90 etcdctl already reads env vars with prefix `ETCDCTL_`. And with this commit, the client library will read vars with prefix `ETCDCLIENT_`. They won't conflict.

**Comment by mitake** [comment]
@xiang90 kindly ping? Or should I consider something related to the prefix?

**Comment by mitake** [comment]
Kindly ping anyone?

**Comment by mitake** [comment]
Any updates?

**Comment by mitake** [comment]
Or is v2 client fixed already?

**Comment by gyuho** [comment]
@mitake Sorry for delay! Do we still want this? We are only fixing bugs in v2 now https://coreos.com/blog/toward-etcd-v3-in-container-linux.html :0

**Comment by mitake** [comment]
@gyuho I just wanted it when I wans doing benchmark of v2 :) So let's close it

**Pull request closed without merging**